lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp

```diff
@@ -552,6 +569,272 @@ class ConvertAtenMaxPool2dWithIndicesBackwardOp
+    if (failed(verifyLinalgCompatibleTypes(op, rewriter)))
+      return failure();
+
+    Location loc = op.getLoc();
+    MLIRContext *context = op->getContext();
+
+    RankedTensorType selfType = adaptor.getSelf().getType().cast<RankedTensorType>();
+    RankedTensorType indexType = adaptor.getIndex().getType().cast<RankedTensorType>();
+    RankedTensorType srcType = adaptor.getSrc().getType().cast<RankedTensorType>();
+
+    Value self = adaptor.getSelf();
+
+    if (selfType.getRank() != indexType.getRank() || indexType.getRank() != srcType.getRank())
+      return rewriter.notifyMatchFailure(
+          op,
+          "self, index and src should all have the same number of dimensions.");
+
+    std::string reduceType;
+    if (!matchPattern(op.getReduce(), m_TorchConstantStr(reduceType)))
+        return rewriter.notifyMatchFailure(op, "'reduce' must be a costant string");
+
+    Value dimValue = op.getDim();
+    int64_t dim;
+    if (!matchPattern(dimValue, m_TorchConstantInt(&dim)))
+        return rewriter.notifyMatchFailure(op, "'dim' is not constant");
+
+    bool includeSelf;
+    if (!matchPattern(op.getIncludeSelf(), m_TorchConstantBool(&includeSelf)))
+        return rewriter.notifyMatchFailure(op,
+                                           "'include_self' is not constant");
+
+    Value indexSize = getTensorSize(rewriter, loc, adaptor.getIndex());
+    indexSize = castIntToIndex(rewriter, loc, indexSize);
+    SmallVector<Value> indexShape = getTensorSizes(rewriter, loc, adaptor.getIndex());
+    Value cstOne = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+
+    // all the indices slices shapes will be (indexSize, 1)
+    SmallVector<Value> indSliceShape({indexSize, cstOne});
+    Value indSlice = createZeroInitTensor(rewriter, loc, indSliceShape,
+                                          rewriter.getI32Type());
+    SmallVector<Value> outputs(indexType.getRank(), indSlice);
+    // updates tensor will be of shape (indexSize).
+    outputs.push_back(createZeroInitTensor(rewriter, loc, {indexSize},
+                                           selfType.getElementType()));
+    SmallVector<Type> outputsType(indexType.getRank(), indSlice.getType());
+    outputsType.push_back(outputs[indexType.getRank()].getType());
+
+    // create mapping
+    SmallVector<AffineExpr> indSliceExpr;
+    indSliceExpr.push_back(rewriter.getAffineDimExpr(0));
+    indSliceExpr.push_back(rewriter.getAffineConstantExpr(0));
+    SmallVector<AffineMap> mapping(
+        indexType.getRank(), AffineMap::get(/*dimCount=*/1, /*symbolCount=*/0,
+                                  indSliceExpr, context));
+    // mapping for updates
+    mapping.push_back(rewriter.getDimIdentityMap());
+    SmallVector<utils::IteratorType> iteratorTypes({utils::IteratorType::parallel});
+
+    // Need to flatten the scatter argument
+    auto tmTensorScatterInputs = rewriter.create<linalg::GenericOp>(
```

Comment:
Potentially refactor into general function so other lowerings can use

Reply:
Yeah, I think extracting the logic that transforms a PyTorch scatter indices tensor to the indices tensor tm_tensor expects can be refactored out into a helper function to make the lowering for this op easier to digest.

Reply:
Done: let me know if there should be any other simplifications